test/SILGen/literals.swift

```diff
@@ -41,3 +41,28 @@ class CustomStringSubclass : CustomStringClass {}
 func returnsCustomStringSubclass() -> CustomStringSubclass {
   return "hello world"
 }
+
+class TakesArrayLiteral<Element> : ExpressibleByArrayLiteral {
+  required init(arrayLiteral elements: Element...) {}
+}
+
+// CHECK-LABEL: sil hidden [ossa] @$s8literals18returnsCustomArrayAA05TakesD7LiteralCySiGyF : $@convention(thin) () -> @owned TakesArrayLiteral<Int>
+// CHECK: [[TMP:%.*]] = apply %2(%0, %1) : $@convention(method) (Builtin.IntLiteral, @thin Int.Type) -> Int
+// CHECK: [[ARRAY_LENGTH:%.*]] = integer_literal $Builtin.Word, 2
+// CHECK: [[ALLOCATE_VARARGS:%.*]] = function_ref @$ss27_allocateUninitializedArrayySayxG_BptBwlF
+// CHECK: [[ARR_TMP:%.*]] = apply [[ALLOCATE_VARARGS]]<Int>([[ARRAY_LENGTH]])
+// CHECK: ([[ARR:%.*]], [[ADDRESS:%.*]]) = destructure_tuple [[ARR_TMP]]
+// CHECK: [[POINTER:%.*]] = pointer_to_address [[ADDRESS]]
+// CHECK: store [[TMP:%.*]] to [trivial] [[POINTER]]
```

Comment:
Can you add tests for Array Literals that take:

1. Classes. E.x.: class Klass {}
2. Address Only Types: struct Foo<T> { var t: T }
3. self
4. Non-Trivial structs: struct Foo { var k: Klass }

That would give me the confidence that we are not doing anything too stupid.

Reply:
Also, mutating/non-mutating self would be good + some interesting lvalue paths from a mutating self. I would also add a test with a protocol.

All of these should show if you really need the scope here/which scopes you need.

Reply:
To be clear, I think you can take the test that you made and replicate it for those types.